File: OffByOne/src/OffByOne.Ql/Ast/Expressions/Binary/Base/BinaryExpression.cs
namespace OffByOne.Ql.Ast.Expressions.Binary.Base
{
    using System;
    using System.Collections.Generic;

    public abstract class BinaryExpression : Expression
    {
        protected BinaryExpression(
            Expression leftExpression,
            Expression rightExpression)
        {
            if (leftExpression == null)
            {
                throw new ArgumentNullException(nameof(leftExpression));
            }

            if (rightExpression == null)
            {
                throw new ArgumentNullException(nameof(rightExpression));
            }

            this.LeftExpression = leftExpression;
            this.RightExpression = rightExpression;
        }

        public Expression LeftExpression { get; }

        public Expression RightExpression { get; }

        public override ISet<string> GetDependencies()
        {
            var identifiers = new HashSet<string>();
            identifiers.UnionWith(this.LeftExpression.GetDependencies());
            identifiers.UnionWith(this.RightExpression.GetDependencies());
            return identifiers;
        }
    }
}

File: AlternativeQL/src/org/ql/ast/expression/BinaryExpression.java
package org.ql.ast.expression;

public abstract class BinaryExpression extends Expression {
    private final Expression left;
    private final Expression right;

    public BinaryExpression(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    public Expression getLeft() {
        return left;
    }

    public Expression getRight() {
        return right;
    }
}
